File: samples/vdsp_op/custom_op/norma_tensor_3ch/norma_tensor_3ch.cpp

/*
 * Copyright (C) 2025 Vastai-tech Company.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include "common/cmdline.hpp"
#include "common/utils.hpp"
#include "norma_tensor_3ch_op.hpp"
#include "vaststreamx/vaststreamx.h"

cmdline::parser ArgumentParser(int argc, char** argv) {
  cmdline::parser args;
  args.add<uint32_t>("device_id", 'd', "device id to run", false, 0);
  args.add<std::string>("elf_file", '\0', "elf file path", false,
                        "/opt/vastai/vaststreamx/data/elf/norma_tensor_3ch");
  args.add<std::string>("input_shape", '\0', "input_shape [c,h,w]", false,
                        "[3,640,640]");
  args.parse_check(argc, argv);
  return args;
}

int main(int argc, char* argv[]) {
  auto args = ArgumentParser(argc, argv);

  uint32_t device_id = args.get<uint32_t>("device_id");
  auto elf_file = args.get<std::string>("elf_file");
  auto input_shape = vsx::ParseShape(args.get<std::string>("input_shape"));

  auto argmax_op =
      vsx::NormaTensor3ChOp("norma_tensor_3ch_op", elf_file, device_id);
  auto input_tensor =
      vsx::Tensor(input_shape, vsx::Context::CPU(), vsx::TypeFlag::kInt8);

  auto output_tensor_vacc = argmax_op.Process(input_tensor);
  auto output_shape = output_tensor_vacc.Shape();
  std::cout << "output tensor shape:" << output_shape << std::endl;
  return 0;
}

File: samples/vdsp_op/custom_op/norma_tensor_3ch/norma_tensor_3ch.py
import os
import sys
import argparse
import numpy as np
from norma_tensor_3ch_op import NormaTensor3ChOp
import ast

current_file_path = os.path.dirname(os.path.abspath(__file__))
common_path = os.path.join(current_file_path, "../../../../")
sys.path.append(common_path)


def argument_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-d",
        "--device_id",
        default=0,
        type=int,
        help="device id to run",
    )
    parser.add_argument(
        "--elf_file",
        default="/opt/vastai/vaststreamx/data/elf/norma_tensor_3ch",
        help="elf file",
    )
    parser.add_argument(
        "-s",
        "--shape",
        help="model input shape",
    )
    args = parser.parse_args()
    return args


if __name__ == "__main__":
    args = argument_parser()

    argmax_op = NormaTensor3ChOp("norma_tensor_3ch_op", args.elf_file, args.device_id)
    shape = ast.literal_eval(args.shape)

    input = np.random.rand(*shape).astype(np.int8)

    output = argmax_op.process(input)

    print("output shape: ", output.shape)
